README.md
<!-- ix-docs-ignore -->
![imgix logo](https://assets.imgix.net/sdk-imgix-logo.svg)

`imgix-java` is a client library for generating image URLs with [imgix](https://www.imgix.com/).

[![Download](https://api.bintray.com/packages/imgix/maven/imgix-java/images/download.svg) ](https://bintray.com/imgix/maven/imgix-java/_latestVersion)
[![Maven Central](https://img.shields.io/maven-central/v/com.imgix/imgix-java.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22com.imgix%22%20AND%20a:%22imgix-java%22)
[![Build Status](https://travis-ci.com/imgix/imgix-java.svg?branch=main)](https://travis-ci.com/imgix/imgix-java)
[![License](https://img.shields.io/github/license/imgix/imgix-java)](https://github.com/imgix/imgix-java/blob/main/LICENSE)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-java.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-java?ref=badge_shield)

---
<!-- /ix-docs-ignore -->

- [Installation](#installation)
    * [Dependencies](#dependencies)
    * [Install Options](#install-options)
        + [Gradle & JCenter](#gradle--jcenter)
        + [Creating a Jar](#creating-a-jar)
- [Usage](#usage)
- [Signed URLs](#signed-urls)
- [Srcset Generation](#srcset-generation)
    * [Fixed-Width Images](#fixed-width-images)
    * [Fluid-Width Images](#fluid-width-images)
        + [Custom Widths](#custom-widths)
        + [Width Ranges](#width-ranges)
        + [Width Tolerance](#width-tolerance)
- [Running Tests](#running-tests)
- [License](#license)

## Installation

### Dependencies

The library itself has no external dependencies. Although if you want to build from source (or run tests) then you need `ant` and the `JDK 1.6+`.

### Install Options

#### Gradle & JCenter

To add Imgix-Java to your project, include the following in your project's build.gradle:

```
dependencies {
   compile "com.imgix:imgix-java:2.3.2"
}
```

And if this is your first external MavenCentral dependency you'll need to add, again to your project level build.gradle, the following:

```
buildscript {
   repositories {
        mavenCentral()
   }
}
```

#### Creating a Jar

To create a jar from source:

```
gradle build
```

This creates `imgix-java-{VERSION_NUMBER}.jar` under `./build/libs`

Once a new version has been merged into main on GitHub (don't forget to update the version numbers in build.gradle first!), it can be deployed to Bintray with `gradle build && gradle bintrayUpload`. After that, the new version can be viewed via the [Bintray web interface](https://bintray.com/imgix/maven/imgix-java).

## Usage

To begin creating imgix URLs, add the jar to your project's classpath and import the imgix library. The URL builder can be reused to create URLs for any images on the domains it is provided.

```java
import com.imgix.URLBuilder;
import java.util.Map;
import java.util.HashMap;

public class ImgixExample {
    public static void main(String[] args) {
        URLBuilder builder = new URLBuilder("demos.imgix.net");
        Map<String, String> params = new HashMap<String, String>();
        params.put("w", "100");
        params.put("h", "100");
        System.out.println(builder.createURL("bridge.png", params));
        // http://demos.imgix.net/bridge.png?h=100&w=100
    }
}
```

HTTPS support is available by default. However, if you need HTTP support, call setUseHttps on the builder:

```java
import com.imgix.URLBuilder;
import java.util.Map;
import java.util.HashMap;

public class ImgixExample {
    public static void main(String[] args) {
        URLBuilder builder = new URLBuilder("demos.imgix.net");

        builder.setUseHttps(false); // use http

        Map<String, String> params = new HashMap<String, String>();
        params.put("w", "100");
        params.put("h", "100");
        System.out.println(builder.createURL("bridge.png", params));
        // https://demos.imgix.net/bridge.png?h=100&w=100
    }
}
```

## Signed URLs

To produce a signed URL, you must enable secure URLs on your source and then provide your signature key to the URL builder.

```java
import com.imgix.URLBuilder;
import java.util.Map;
import java.util.HashMap;

public class ImgixExample {
    public static void main(String[] args) {
        URLBuilder builder = new URLBuilder("demos.imgix.net");
        builder.setSignKey("test1234"); // set sign key
        Map<String, String> params = new HashMap<String, String>();
        params.put("w", "100");
        params.put("h", "100");
        System.out.println(builder.createURL("bridge.png", params));
        // http://demos.imgix.net/bridge.png?h=100&w=100&s=bb8f3a2ab832e35997456823272103a4
    }
}
```

## Srcset Generation

The imgix-java library allows for generation of custom `srcset` attributes, which can be invoked through `createSrcSet()`. By default, the `srcset` generated will allow for responsive size switching by building a list of image-width mappings.

```java
URLBuilder ub = new URLBuilder("demos.imgix.net", true, "my-token", false);
String srcset = ub.createSrcSet("bridge.png");
System.out.println(srcset);
```
The above will produce the following srcset attribute value which can then be served to the client:

```html
https://demos.imgix.net/bridge.png?w=100&s=494158d968e94ac8e83772ada9a83ad1 100w,
https://demos.imgix.net/bridge.png?w=116&s=6a22236e189b6a9548b531330647ffa7 116w,
https://demos.imgix.net/bridge.png?w=135&s=cbf91f556dd67c0b9e26cb9784a83794 135w,
                                    ...
https://demos.imgix.net/bridge.png?w=7401&s=503e3ba04588f1c301863c9a5d84fe91 7401w,
https://demos.imgix.net/bridge.png?w=8192&s=152551ce4ec155f7a03f60f762a1ca33 8192w
```

### Fixed-Width Images

In cases where enough information is provided about an image's dimensions, `createSrcSet()` will instead build a `srcset` that will allow for an image to be served at different resolutions. The parameters taken into consideration when determining if an image is fixed-width are `w` (width) and `h` (height).

By invoking `createSrcSet()` with either a width **or** height provided, a different `srcset` will be generated for a fixed-size image instead.

```java
URLBuilder ub = new URLBuilder("demos.imgix.net", true, "my-token", false);
HashMap<String,String> params = new HashMap<String,String> ();
params.put("h", "200");
params.put("ar", "3:2");
params.put("fit", "crop");
String srcset = ub.createSrcSet("bridge.png", params);
System.out.println(srcset);

```

Will produce the following attribute value:

```html
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=1&fit=crop&h=200&s=4c79373f535df7e2594a8f6622ec6631 1x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=2&fit=crop&h=200&s=dc818ae4522494f2f750651304a4d825 2x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=3&fit=crop&h=200&s=ba1ec0cef6c77ff02330d40cc4dae932 3x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=4&fit=crop&h=200&s=b51e497d9461be62354c0ea12b6524fb 4x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=5&fit=crop&h=200&s=dc37c1fbee505d425ca8e3764b37f791 5x
```

For more information to better understand `srcset`, we recommend [Eric Portis' "Srcset and sizes" article](https://ericportis.com/posts/2014/srcset-sizes/) which goes into depth about the subject.


##### Variable Quality

This library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset. This technique is commonly used to compensate for the increased file size of high-DPR images.

Since high-DPR images are displayed at a higher pixel density on devices, image quality can be lowered to reduce overall file size without sacrificing perceived visual quality. For more information and examples of this technique in action, see [this blog post](https://blog.imgix.com/2016/03/30/dpr-quality).

This behavior will respect any overriding `q` value passed in as a parameter. Additionally, it can be disabled altogether by passing `disableVariableQuality = true` to `createSrcSet`.

This behavior specifically occurs when a [fixed-width image](#fixed-width-images) is rendered, for example:

```java
URLBuilder ub = new URLBuilder("demo.imgix.net", true, "", false);
HashMap<String, String>  params = new HashMap<String, String>();
params.put("w", "100");
String actual = ub.createSrcSet("image.jpg", params, false);
```

The above will generate a srcset with the following `q` to `dpr` query `params`:

```html
https://demo.imgix.net/image.jpg?dpr=1&q=75&w=100 1x,
https://demo.imgix.net/image.jpg?dpr=2&q=50&w=100 2x,
https://demo.imgix.net/image.jpg?dpr=3&q=35&w=100 3x,
https://demo.imgix.net/image.jpg?dpr=4&q=23&w=100 4x,
https://demo.imgix.net/image.jpg?dpr=5&q=20&w=100 5x
```

### Fluid-Width Images

#### Custom Widths
In situations where specific widths are desired when generating `srcset` pairs, a user can specify them by passing an array of positive integers as `widths`:

``` java
URLBuilder ub = new URLBuilder("demo.imgix.net", true, "", false);
HashMap<String, String>  params = new HashMap<String, String>();
Integer[] widths = new Integer[] {144, 240, 320, 446, 640};

String srcset = ub.createSrcSet("image.jpg", params, widths);
```

```html
https://demo.imgix.net/image.jpg?w=144 144w,
https://demo.imgix.net/image.jpg?w=240 240w,
https://demo.imgix.net/image.jpg?w=320 320w,
https://demo.imgix.net/image.jpg?w=446 446w,
https://demo.imgix.net/image.jpg?w=640 640w
```

**Note**: in situations where a `srcset` is being rendered as a [fixed image](#fixed-image-rendering), any custom `widths` passed in will be ignored.

Additionally, if both `widths` and a width `tol`erance are passed to the `createSrcSet` method, the custom widths list will take precedence.

#### Width Ranges

In certain circumstances, you may want to limit the minimum or maximum value of the non-fixed (fluid-width) `srcset` generated by the `createSrcSet` method. To do this, you can specify the widths at which a srcset should `begin` and `end`:

```java
URLBuilder ub = new URLBuilder("demo.imgix.net", true, "", false);
HashMap<String, String>  params = new HashMap<String, String>();
String actual = ub.createSrcSet("image.jpg", params, 500, 2000);
```

Formatted version of the above srcset attribute:

``` html
https://demo.imgix.net/image.jpg?w=500 500w,
https://demo.imgix.net/image.jpg?w=580 580w,
https://demo.imgix.net/image.jpg?w=673 673w,
https://demo.imgix.net/image.jpg?w=780 780w,
https://demo.imgix.net/image.jpg?w=905 905w,
https://demo.imgix.net/image.jpg?w=1050 1050w,
https://demo.imgix.net/image.jpg?w=1218 1218w,
https://demo.imgix.net/image.jpg?w=1413 1413w,
https://demo.imgix.net/image.jpg?w=1639 1639w,
https://demo.imgix.net/image.jpg?w=1901 1901w,
https://demo.imgix.net/image.jpg?w=2000 2000w
```

#### Width Tolerance

The `srcset` width `tol`erance dictates the maximum `tol`erated difference between an image's downloaded size and its rendered size.

For example, setting this value to 0.1 means that an image will not render more than 10% larger or smaller than its native size. In practice, the image URLs generated for a width-based srcset attribute will grow by twice this rate.

A lower tolerance means images will render closer to their native size (thereby increasing perceived image quality), but a large srcset list will be generated and consequently users may experience lower rates of cache-hit for pre-rendered images on your site.

By default, srcset width `tol`erance is set to `0.08` (8 percent), which we consider to be the ideal rate for maximizing cache hits without sacrificing visual quality. Users can specify their own width tolerance by providing a positive scalar value as width `tol`erance:

```java
URLBuilder ub = new URLBuilder("test.imgix.net", false, "", false);
HashMap<String, String>  params = new HashMap<String, String>();
String srcset = ub.createSrcSet("image.png", params, 100, 384, 0.20);
```

In this case, the width `tol`erance is set to 20 percent, which will be reflected in the difference between subsequent widths in a srcset pair:

```html
https://demo.imgix.net/image.jpg?w=100 100w,
https://demo.imgix.net/image.jpg?w=140 140w,
https://demo.imgix.net/image.jpg?w=196 196w,
https://demo.imgix.net/image.jpg?w=274 274w,
https://demo.imgix.net/image.jpg?w=384 384w
```

## Running Tests

To run tests clone this project and run:

```
gradle test
```

Dependencies for running tests (junit, etc) are provided (in `test/lib` and referenced in the build config).

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-java.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-java?ref=badge_large)


CHANGELOG.md
Change Log
==========

### Version 2.3.2 (_April 13, 2021_)
* fix(sanitizePath): ensure unicode chars encoded ([#70](https://github.com/imgix/imgix-java/pull/70))
* fix(URLHelper): add sanitizePath method ([#68](https://github.com/imgix/imgix-java/pull/68))

### Version 2.3.1 (_March 24, 2021_)
* docs: add maven badge to README.md ([#65](https://github.com/imgix/imgix-java/pull/65))
* docs: update fixed-widths section ([#63](https://github.com/imgix/imgix-java/pull/63))
* fix: create dpr srcset when only h passed in as param ([#62](https://github.com/imgix/imgix-java/pull/62))
* docs: update travis badge to travis-ci.com ([#60](https://github.com/imgix/imgix-java/pull/60))
* fix(gradle-build): include maven plugin ([#55](https://github.com/imgix/imgix-java/pull/55))
* docs(readme): correct percentage typos ([#53](https://github.com/imgix/imgix-java/pull/53))

### Version 2.3.0 (_June 05, 2020_)
* feat: custom srcsets ([#44](https://github.com/imgix/imgix-java/pull/44))
* fix: convert tol to float ([#49](https://github.com/imgix/imgix-java/pull/49))
* feat: throw when validator detects invalid input ([#52](https://github.com/imgix/imgix-java/pull/52))

### Version 2.2.0 (_April 10, 2020_)

* feat(https): use https by default ([#38](https://github.com/imgix/imgix-java/pull/38))
* docs(readme): update to reflect https by default ([#39](https://github.com/imgix/imgix-java/pull/39))
* feat: prevent push mistakes by updating .gitignore ([#41](https://github.com/imgix/imgix-java/pull/41))
* build: upgrade gradle usage to prep for 7.0 ([#42](https://github.com/imgix/imgix-java/pull/42))

### Version 2.1.1 (_December 05, 2019_)

* fix: explicitly convert string to UTF-8 byte array ([#36](https://github.com/imgix/imgix-java/pull/36))

## Version 2.1.0 (_September 19, 2019_)

* fix: add domain validation during URLBuilder initialization ([#33](https://github.com/imgix/imgix-java/pull/33))
* feat: add srcset generation method ([#32](https://github.com/imgix/imgix-java/pull/32)) 

# Version 2.0.0 (_May 17, 2019_)

* fix: remove deprecated domain sharding functionality  ([#31](https://github.com/imgix/imgix-java/pull/31))
* refactor: rename `signWithLibraryParameter` to `includeLibraryParam` ([#30](https://github.com/imgix/imgix-java/pull/30))

## Version 1.2.0 (_May 17, 2019_)

* fix: deprecate domain sharding ([#29](https://github.com/imgix/imgix-java/pull/29))

### Version 1.1.12 (_Jul 19, 2018_)

* Support Java 10

### Version 1.1.10 (_Jun 15, 2017_)

* Adding javadoc and source jars to the build artifacts.

### Version 1.1.9 (_Jun 15, 2017_)

* No longer URL encode path components, which prevented signed URLs from working (and was also just wrong).

### Version 1.1.8 (_Sep 13, 2016_)

* Finally released to JCenter! https://bintray.com/imgix/maven/imgix-java

### Version 1.1.7 (_Jul 22, 2016_)

### Version 1.1.5 (_Jul 22, 2016_)

### Version 1.1.4 (_Jul 22, 2016_)

### Version 1.1.2 (_Jul 15, 2016_)

### Version 1.1.0 (_Feb 25, 2016_)

* Added automatic Base64 encoding for all Base64 variant parameters.
* Properly encoding all query string keys and values.


src/main/java/com/imgix/URLBuilder.java
package com.imgix;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.regex.Pattern;

public class URLBuilder {

  public static final String VERSION = "2.3.2";
  private static final String DOMAIN_REGEX =
      "^(?:[a-z\\d\\-_]{1,62}\\.){0,125}(?:[a-z\\d](?:\\-(?=\\-*[a-z\\d])|[a-z]|\\d){0,62}\\.)[a-z\\d]{1,63}$";

  private String domain;
  private boolean useHttps;
  private String signKey;
  private boolean includeLibraryParam;

  private static final Integer[] SRCSET_TARGET_WIDTHS = {
    100, 116, 135, 156, 181, 210, 244, 283, 328, 380, 441, 512, 594, 689, 799, 927, 1075, 1247,
    1446, 1678, 1946, 2257, 2619, 3038, 3524, 4087, 4741, 5500, 6380, 7401, 8192
  };

  private static final double SRCSET_WIDTH_TOLERANCE = 0.08;
  private static final int MIN_WIDTH = 100;
  private static final int MAX_WIDTH = 8192;
  private static final Integer[] DPR_QUALITIES = {75, 50, 35, 23, 20};
  private static final Integer[] TARGET_RATIOS = {1, 2, 3, 4, 5};

  public URLBuilder(String domain, boolean useHttps, String signKey, boolean includeLibraryParam) {
    Pattern domainPattern = Pattern.compile(DOMAIN_REGEX);

    if (domain == null || domain.length() == 0) {
      throw new IllegalArgumentException("At lease one domain must be passed to URLBuilder");
    } else if (!domainPattern.matcher(domain).matches()) {
      throw new IllegalArgumentException(
          "Domain must be passed in as a fully-qualified domain name and should not include a"
              + " protocol or any path element, i.e. \"example.imgix.net\".");
    }

    this.domain = domain;
    this.useHttps = useHttps;
    this.signKey = signKey;
    this.includeLibraryParam = includeLibraryParam;
  }

  /**
   * This single-parameter constructor, accepts a domain string and returns a `URLBuilder` with it's
   * `useHttps` member variable set to `true` by default.
   *
   * @param domain - a valid domain string, e.g. `example.imgix.net`
   */
  public URLBuilder(String domain) {
    this(domain, true);
  }

  public URLBuilder(String domain, boolean useHttps) {
    this(domain, useHttps, "");
  }

  public URLBuilder(String domain, boolean useHttps, String signKey) {
    this(domain, useHttps, signKey, true);
  }

  public void setUseHttps(boolean useHttps) {
    this.useHttps = useHttps;
  }

  public void setSignKey(String signKey) {
    this.signKey = signKey;
  }

  public String createURL(String path) {
    return createURL(path, new TreeMap<String, String>());
  }

  public String createURL(String path, Map<String, String> params) {
    String scheme = this.useHttps ? "https" : "http";

    if (this.includeLibraryParam) {
      params.put("ixlib", "java-" + VERSION);
    }

    return new URLHelper(domain, path, scheme, signKey, params).getURL();
  }

  /**
   * Create a srcset given a `path` and a map of `params`.
   *
   * <p>This function creates a dpr based srcset if `params` contain either: - a width "w" param
   * _or_ a height "h" param.
   *
   * <p>Otherwise, a srcset of width-pairs is created.
   *
   * @param path - path to the image, i.e. "image/file.png"
   * @param params - map of query parameters
   * @return srcset attribute string
   */
  public String createSrcSet(String path, Map<String, String> params) {
    return createSrcSet(path, params, MIN_WIDTH, MAX_WIDTH, SRCSET_WIDTH_TOLERANCE, false);
  }

  public String createSrcSet(String path) {
    return createSrcSet(path, new TreeMap<String, String>());
  }

  /**
   * Create a srcset given a `path`, map of `params`, and `tol`.
   *
   * <p>This function provides the ability to vary `tol` while generating a srcset over the default
   * range.
   *
   * @param path - path to the image, i.e. "image/file.png"
   * @param params - map of query parameters
   * @param tol - tolerable amount of width value variation
   * @return srcset attribute string
   */
  public String createSrcSet(String path, Map<String, String> params, double tol) {
    return createSrcSet(path, params, MIN_WIDTH, MAX_WIDTH, tol, false);
  }

  /**
   * Create a srcset given a `path`, map of `params`, `begin` and `end`.
   *
   * <p>This function provides the ability to create a srcset attribute over a custom range between
   * `begin` and `end` (inclusively) while using the default srcset width tolerance.
   *
   * @param path - path to the image, i.e. "image/file.png"
   * @param params - map of query parameters
   * @param begin - beginning image width value
   * @param end - ending image width value
   * @return srcset attribute string
   */
  public String createSrcSet(String path, Map<String, String> params, int begin, int end) {
    return createSrcSet(path, params, begin, end, SRCSET_WIDTH_TOLERANCE, false);
  }

  /**
   * Create a srcset given a `path`, map of `params`, `begin`, `end`, and `tol`.
   *
   * <p>This function provides the ability to create a completely custom srcset attribute. The
   * srcset width value range will `begin` and `end` on the specified values. The specified `tol`
   * determines the tolerable amount of width value variation.
   *
   * @param path - path to the image, i.e. "image/file.png"
   * @param params - map of query parameters
   * @param begin - beginning image width value
   * @param end - ending image width value
   * @param tol - tolerable amount of width value variation, from 0.01 to 1.0
   * @return srcset attribute string
   */
  public String createSrcSet(
      String path, Map<String, String> params, int begin, int end, double tol) {
    return createSrcSet(path, params, begin, end, tol, false);
  }

  /**
   * Create a srcset given a `path`, map of `params`, and the `disableVariableQuality` flag.
   *
   * <p>If `disableVariableQuality` is `false` then variable output is turned _on_. If
   * `disableVariableQuality` is `true` then variable quality output is turned _off_.
   *
   * @param path - path to the image, i.e. "image/file.png"
   * @param params - map of query parameters
   * @param disableVariableQuality - flag to toggle variable image output quality.
   * @return srcset attribute string
   */
  public String createSrcSet(
      String path, Map<String, String> params, boolean disableVariableQuality) {
    return createSrcSet(
        path, params, MIN_WIDTH, MAX_WIDTH, SRCSET_WIDTH_TOLERANCE, disableVariableQuality);
  }

  public String createSrcSet(
      String path,
      Map<String, String> params,
      int begin,
      int end,
      double tol,
      boolean disableVariableQuality) {
    if (isDpr(params)) {
      return createSrcSetDPR(path, params, disableVariableQuality);
    } else {
      Integer[] targets = targetWidths(begin, end, tol).toArray(new Integer[0]);
      return createSrcSetPairs(path, params, targets);
    }
  }

  public String createSrcSet(String path, HashMap<String, String> params, Integer[] widths) {
    return createSrcSetPairs(path, params, widths);
  }

  private String createSrcSetPairs(String path, Map<String, String> params) {
    return createSrcSetPairs(path, params, SRCSET_TARGET_WIDTHS);
  }

  private String createSrcSetPairs(String path, Map<String, String> params, Integer[] widths) {
    Validator.validateWidths(widths);

    StringBuilder srcset = new StringBuilder();

    for (Integer width : widths) {
      params.put("w", width.toString());
      srcset.append(this.createURL(path, params)).append(" ").append(width).append("w,\n");
    }

    return srcset.substring(0, srcset.length() - 2);
  }

  private String createSrcSetDPR(String path, Map<String, String> params) {
    return createSrcSetDPR(path, params, false);
  }

  private String createSrcSetDPR(
      String path, Map<String, String> params, boolean disableVariableQuality) {
    StringBuilder srcset = new StringBuilder();
    Map<String, String> srcsetParams = new HashMap<String, String>(params);

    boolean has_quality = params.get("q") != null;

    for (int ratio : TARGET_RATIOS) {
      srcsetParams.put("dpr", Integer.toString(ratio));

      if (!disableVariableQuality && !has_quality) {
        srcsetParams.put("q", DPR_QUALITIES[ratio - 1].toString());
      }
      srcset.append(this.createURL(path, srcsetParams)).append(" ").append(ratio).append("x,\n");
    }
    return srcset.substring(0, srcset.length() - 2);
  }

  /**
   * Create an `ArrayList` of integer target widths.
   *
   * <p>If `begin`, `end`, and `tol` are the default values, then the targets are not custom, in
   * which case the default widths are returned.
   *
   * <p>A target width list of length one is valid: if `begin` == `end` then the list begins where
   * it ends.
   *
   * <p>When the `while` loop terminates, `begin` is greater than `end` (where `end` less than or
   * equal to `MAX_WIDTH`). This means that the most recently appended value may, or may not, be the
   * `end` value.
   *
   * <p>To be inclusive of the ending value, we check for this case and the value is added if
   * necessary.
   *
   * @param begin - beginning image width value
   * @param end - ending image width value
   * @param tol - tolerable amount of width value variation, from 0.01 to 1.0
   * @return array list of image width values
   */
  public static ArrayList<Integer> targetWidths(int begin, int end, double tol) {
    Validator.validateMinMaxTol(begin, end, tol);
    return computeTargetWidths(begin, end, tol);
  }

  public static ArrayList<Integer> targetWidths() {
    return new ArrayList<Integer>(Arrays.asList(SRCSET_TARGET_WIDTHS));
  }

  /**
   * Create an `ArrayList` of integer target widths.
   *
   * <p>This function is the implementation details of `targetWidths`. This function exists to
   * provide a consistent interface for callers of `targetWidths`.
   *
   * <p>This function implements the syntax that fulfills the semantics of `targetWidths`. Meaning,
   * `begin`, `end`, and `tol` are to be whole integers, but computation requires `double`s. This
   * function hides this detail from callers.
   */
  private static ArrayList<Integer> computeTargetWidths(double begin, double end, double tol) {
    if (notCustom(begin, end, tol)) {
      return new ArrayList<Integer>(Arrays.asList(SRCSET_TARGET_WIDTHS));
    }

    ArrayList<Integer> resolutions = new ArrayList<Integer>();
    if (begin == end) {
      // `begin` has not been mutated; cast back to `int`.
      resolutions.add((int) begin);
      return resolutions;
    }

    while (begin < end && begin < MAX_WIDTH) {
      // Round values so that the resulting `int` is truer
      // to expectations (i.e. 115.99999 --> 116).
      resolutions.add((int) Math.round(begin));
      begin *= 1 + tol * 2;
    }

    int lastIndex = resolutions.size() - 1;
    if (resolutions.get(lastIndex) < end) {
      // `end` has not been mutated; cast back to `int`.
      resolutions.add((int) end);
    }

    return resolutions;
  }

  private boolean isDpr(Map<String, String> params) {
    String width = params.get("w");
    boolean hasWidth = (width != null) && !width.isEmpty();

    String height = params.get("h");
    boolean hasHeight = (height != null) && !height.isEmpty();

    String aspectRatio = params.get("ar");
    boolean hasAspectRatio = (aspectRatio != null) && !aspectRatio.isEmpty();

    // If `params` have a width param or height parameters
    // then the srcset to be constructed with these params
    // _is dpr based_.
    return hasWidth || hasHeight;
  }

  private static boolean notCustom(double begin, double end, double tol) {
    boolean defaultBegin = (begin == MIN_WIDTH);
    boolean defaultEnd = (end == MAX_WIDTH);
    boolean defaultTol = (tol == SRCSET_WIDTH_TOLERANCE);

    // A list of target widths is _NOT_ custom if `begin`, `end`,
    // and `tol` are equal to their default values.
    return defaultBegin && defaultEnd && defaultTol;
  }
}
